### PR TITLE
Add uhd

### PR DESCRIPTION
**Pull request by ryanvolz**
UHD is a C++ library with Python bindings for interfacing with USRP software-defined radios produced by Ettus Research. It's a fairly important dependency of GNU Radio, which is widely used in the software radio community. (A recipe for GNU Radio will follow once recipes for a few of its dependencies are merged.)

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by ryanvolz** [comment]
@conda-forge/help-python-c This is ready for review. The AppVeyor timeout is expected, and we can disable it in favor of Azure in the feedstock.

**Comment by ryanvolz** [comment]
cc @conda-forge/staged-recipes This is ready for review. Thanks!

File: recipes/uhd/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    - mako >=0.5.0
+    - ncurses  # [not win]
+    - numpy
+    - python
+    - requests
+    - setuptools
+
+  run:
+    - boost
+    - boost-cpp
```
Comment (chrisburr) [inline recipes/uhd/meta.yaml]: I think this only contains headers so you won’t need it at runtime?

Comment (beckermr) [reply recipes/uhd/meta.yaml]: It looks like boost-cpp is a dep of boost so it gets pulled in no matter what.

https://github.com/conda-forge/boost-feedstock/blob/master/recipe/meta.yaml#L28

Only actual deps of `uhd` should be listed. Any deps of deps can be excluded.

File: recipes/uhd/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  skip: True  # [win and py2k]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake >=3.5.1
+    - jom  # [win]
+    - {{ cdt('libudev-devel') }}  # [linux]
+    - pkg-config  # [not win]
```
[the remark this answers is not in the document]
Comment (ryanvolz) [reply recipes/uhd/meta.yaml]: Yes, good catch!

File: recipes/uhd/meta.yaml (hunk 1 above)
Comment (ryanvolz) [reply recipes/uhd/meta.yaml]: They are both definitely needed in host for the build (although the Boost Python requirement should be going away with the next release), and the info after the build spits out a lot of "needed DSO" pertaining to ``boost-cpp`` and one pertaining to ``boost``. So I think both are directly required as a run dependency of ``uhd``. Please advise if not though.

Comment (beckermr) [reply recipes/uhd/meta.yaml]: Sure sounds good to me! As long as we are not pinning the versions and letting the global pins take care of it, there is no harm in keeping them. The conda build warnings are of mixed use in my experience. Many codes overlink and so I tend to ignore things but maybe I shouldn't!

Comment (ryanvolz) [reply recipes/uhd/meta.yaml]: Ah, good for me to keep in mind. But yes, in this case I think it's correct.

**Comment by ryanvolz** [comment]
@conda-forge/staged-recipes Gotten some review, ready to merge?

I have also now submitted the relevant (non-backport) patches upstream (https://github.com/EttusResearch/uhd/pull/309, https://github.com/EttusResearch/uhd/pull/310).

File: recipes/uhd/0002-python-Fix-internal-library-name-incl.-suffix-to-mat.patch [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
++++ b/host/python/CMakeLists.txt
+@@ -23,6 +23,12 @@ execute_process(
+ 
+ # Build pyuhd library
+ add_library(pyuhd SHARED pyuhd.cpp)
++# python expects extension modules with a particular suffix
++if(WIN32)
++    set_target_properties(pyuhd PROPERTIES PREFIX "lib" SUFFIX ".pyd")
++else()
++    set_target_properties(pyuhd PROPERTIES PREFIX "lib" SUFFIX ".so")
```
Comment (isuruf) [inline recipes/uhd/0002-python-Fix-internal-library-name-incl.-suffix-to-mat.patch]: FYI, you should use the EXT_SUFFIX value described in PEP 3149, but not necessary to get this PR merged. See https://github.com/symengine/symengine.py/commit/fc5965fa1b34cba4266cc7b096161c5fa81c09b2 for an example.

File: recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch [hunk 4]
```diff
@@ -0,0 +1,12 @@
+diff --git a/host/CMakeLists.txt b/host/CMakeLists.txt
+index 916138ee9..0bce6c726 100644
+--- a/host/CMakeLists.txt
++++ b/host/CMakeLists.txt
+@@ -260,6 +260,7 @@ if(MSVC)
+         -D_CRT_NONSTDC_NO_WARNINGS
+         -D_CRT_NONSTDC_NO_DEPRECATE
+         -D_WINSOCK_DEPRECATED_NO_WARNINGS
++        -DHAVE_SNPRINTF
```
Comment (isuruf) [inline recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch]: You can do `set "CPPFLAGS=%CPPFLAGS% -DHAVE_SNPRINTF" in bld.bat and avoid this patch.

File: recipes/uhd/0002-python-Fix-internal-library-name-incl.-suffix-to-mat.patch (hunk 3 above)
Comment (ryanvolz) [reply recipes/uhd/0002-python-Fix-internal-library-name-incl.-suffix-to-mat.patch]: Great, good to know! I'll incorporate that in the patch I've sent upstream.

File: recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch (hunk 4 above)
Comment (ryanvolz) [reply recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch]: Don't know why I didn't try that in the first place, but yeah that's much easier.

File: recipes/uhd/bld.bat [hunk 5]
```diff
@@ -0,0 +1,69 @@
+setlocal EnableDelayedExpansion
+
+:: prevent redefining snprintf to _snprintf and build failing
+set CPPFLAGS="%CPPFLAGS% -DHAVE_SNPRINTF"
```
Comment (isuruf) [inline recipes/uhd/bld.bat]: windows is weird.
```suggestion
set "CPPFLAGS=%CPPFLAGS% -DHAVE_SNPRINTF"
```

File: recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch (hunk 4 above)
Comment (ryanvolz) [reply recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch]: Looks like CMake is not picking up ``-DHAVE_SNPRINTF`` from the ``CPPFLAGS`` variable. Any pointers @isuruf?

Comment (isuruf) [reply recipes/uhd/0001-Fix-snprintf-redeclared-as_snprintf.patch]: Looks like you need `CXXFLAGS` and `CFLAGS` as CMake ignores `CPPFLAGS`

**Comment by ryanvolz** [comment]
Looks good to go! The only CI failure now is a timeout on AppVeyor. Thanks @chrisburr, @beckermr, @isuruf!

**Pull request merged**